### PR TITLE
speed up docker build

### LINKED ISSUE #1729: Improve `hack/local-up-karmada.sh` execution time
**What would you like to be added**:
The `hack/local-up-karmada.sh` is used for users and developers to set up the Karmada environment quickly. 
But the execution time suddenly gets very long, especially after we change the way building images in #927.

```
# hack/local-up-karmada.sh
...
real	14m12.453s
user	0m28.787s
sys	0m13.848
```
Now it takes about `14` minutes on my side and it just needs `3~5` minutes before.

We can get the execution time change from CI logs as well(from the job `setup e2e test environment`):
- Before #927, it takes about `8` minutes.  [click me for full logs](https://github.com/karmada-io/karmada/runs/6190287719?check_suite_focus=true#step:4:1)
- After #927, it takes about `32` minutes. [click me for full logs](https://github.com/karmada-io/karmada/runs/6203898128?check_suite_focus=true#step:4:1)

**Why is this needed**:
This change has affected users to evaluate Karmada. See #1719

cc @gy95 @ikaven1024 
(I'm not blaming you, just want to notice that we might have to make changes based on your work.)
/help

### PR DESCRIPTION
**Pull request by ikaven1024**
Signed-off-by: yingjinhui <yingjinhui@didiglobal.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Speed up docker image building.

**Which issue(s) this PR fixes**:
Fixes #1729

**Special notes for your reviewer**:
Implement of https://github.com/karmada-io/karmada/issues/1729#issuecomment-1120238596.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by ikaven1024** [comment]
10min [click me for full logs](https://github.com/karmada-io/karmada/runs/6341508072?check_suite_focus=true#step:4:1)
@RainbowMango

**Comment by RainbowMango** [comment]
Thanks @ikaven1024 , I see there is a `WIP` in the PR, does this ready for review?

**Comment by ikaven1024** [comment]
> Thanks @ikaven1024 , I see there is a `WIP` in the PR, does this ready for review?

You can review it now. Some remaining work is update doc and remove debugging code.

**Comment by RainbowMango** [comment]
I see it fetching images from `fetch http://mirrors.ustc.edu.cn/alpine/v3.15/main/x86_64/APKINDEX.tar.gz`, why fetching images from the mirror? Is it due to #1656 ?

**Comment by ikaven1024** [comment]
It works well yesterday all the day [successful log at 5.8](https://github.com/karmada-io/karmada/runs/6341508037?check_suite_focus=true#step:4:81), also fetch from mirrors.ustc.edu.cn.

But failures appear after 12 clock PM. 
https://github.com/karmada-io/karmada/runs/6341953279?check_suite_focus=true#step:4:286
https://github.com/karmada-io/karmada/runs/6342037726?check_suite_focus=true#step:4:51

**Comment by ikaven1024** [comment]
> I see it fetching images from `fetch http://mirrors.ustc.edu.cn/alpine/v3.15/main/x86_64/APKINDEX.tar.gz`, why fetching images from the mirror? Is it due to #1656 ?
> 
> https://github.com/karmada-io/karmada/runs/6342037726?check_suite_focus=true#step:4:50

Ping is forbidden in k8s.gcr.io? I cannot test it, as having no machine out of china mainland. 

In #1656, fetching from `mirrors.ustc.edu.cn`. [see log](https://github.com/karmada-io/karmada/runs/6212705400?check_suite_focus=true#step:4:107)

**Comment by RainbowMango** [comment]
Yeah, I made a test by [this commit](https://github.com/RainbowMango/karmada/commit/9b261ac669f6295c76bf6cf300abfba0731a6352). And found that `ping` not work on GitHub runners:
```
PING googlecode.l.googleusercontent.com (172.253.115.[8](https://github.com/RainbowMango/karmada/runs/6345765429?check_suite_focus=true#step:4:8)2) 56(84) bytes of data.

--- googlecode.l.googleusercontent.com ping statistics ---
3 packets transmitted, 0 received, [10](https://github.com/RainbowMango/karmada/runs/6345765429?check_suite_focus=true#step:4:10)0% packet loss, time 406ms
```
For full logs, you can refer to https://github.com/RainbowMango/karmada/runs/6345765429?check_suite_focus=true#step:4:10.

I don't know why, might be the security concerns. Should we revert #1656?  I don't want CI load images from a mirror, that's not safe and might slow down the test process(Load image from China?).

How do you think? @my-git9

**Comment by my-git9** [comment]
> Yeah, I made a test by [this commit](https://github.com/RainbowMango/karmada/commit/9b261ac669f6295c76bf6cf300abfba0731a6352). And found that `ping` not work on GitHub runners:
> 
> ```
> PING googlecode.l.googleusercontent.com (172.253.115.[8](https://github.com/RainbowMango/karmada/runs/6345765429?check_suite_focus=true#step:4:8)2) 56(84) bytes of data.
> 
> --- googlecode.l.googleusercontent.com ping statistics ---
> 3 packets transmitted, 0 received, [10](https://github.com/RainbowMango/karmada/runs/6345765429?check_suite_focus=true#step:4:10)0% packet loss, time 406ms
> ```
> 
> For full logs, you can refer to https://github.com/RainbowMango/karmada/runs/6345765429?check_suite_focus=true#step:4:10.
> 
> I don't know why, might be the security concerns. Should we revert #1656? I don't want CI load images from a mirror, that's not safe and might slow down the test process(Load image from China?).
> 
> How do you think? @my-git9

LGTM

**Comment by ikaven1024** [comment]
@RainbowMango Ready for merge

File: hack/docker.sh [hunk 1]
```diff
@@ -5,6 +5,21 @@ set -o nounset
...
+# Args:
+#   $1:              target to build
+#   BUILD_PLATFORMS: platforms to build. You can set one or more platforms separated by comma.
+#                    e.g.: linux/amd64,linux/amd64
+#   OUTPUT_TYPE      Destination to save image(docker/registry).
+#   REGISTRY         image registry
+#   VERSION          image version
+# Examples:
+#   hack/docker.sh karmada-aggregated-apiserver
+#   BUILD_PLATFORMS=linux/amd64,linux/amd64 hack/docker.sh karmada-aggregated-apiserver
```
Comment (gy95) [inline hack/docker.sh:21]:
```
# BUILD_PLATFORMS=linux/amd64,linux/amd64 hack/docker.sh karmada-aggregated-apiserver
Cross building image for linux/amd64,linux/amd64: swr.ap-southeast-1.myhuaweicloud.com/karmada/karmada-aggregated-apiserver:unknown
+ docker buildx build --output=type=docker --platform linux/amd64,linux/amd64 --build-arg BINARY=karmada-aggregated-apiserver --tag swr.ap-southeast-1.myhuaweicloud.com/karmada/karmada-aggregated-apiserver:unknown --file hack/../cluster/images/buildx.Dockerfile hack/../_output/bin
[+] Building 0.0s (0/0)                                                                                                                                                      
error: docker exporter does not currently support exporting manifest lists
```

File: .github/workflows/dockerhub-latest-image.yml [hunk 2]
```diff
@@ -30,6 +30,10 @@ jobs:
           # 0 indicates all history for all branches and tags.
           # for `git describe --tags` in Makefile.
           fetch-depth: 0
+      - name: install Go
+        uses: actions/setup-go@v2
+        with:
+          go-version: 1.17.x
```
Comment (gy95) [inline .github/workflows/dockerhub-latest-image.yml:36]: line 49, we need to fix also.

File: .github/workflows/dockerhub-released-image.yml [hunk 3]
```diff
@@ -26,6 +26,10 @@ jobs:
           # 0 indicates all history for all branches and tags.
           # for `git describe --tags` in Makefile.
           fetch-depth: 0
+      - name: install Go
+        uses: actions/setup-go@v2
+        with:
+          go-version: 1.17.x
```
Comment (gy95) [inline .github/workflows/dockerhub-released-image.yml:32]: line 45 `run: make mp-image-${{ matrix.target }}` need to be fixed.

File: hack/docker.sh (hunk 1 above)
Comment (ikaven1024) [reply hack/docker.sh:21]: em, it's my fault. We cannot buildx multi platform with `--output=type=docker`. As this script is not faced to user(user shall use `make image` or `make mp`), so i didn't add validate here. I will change this doc.

File: .github/workflows/dockerhub-released-image.yml (hunk 3 above)
Comment (ikaven1024) [reply .github/workflows/dockerhub-released-image.yml:32]: What's wrong with it? Sorry, i cannot get it...

Comment (gy95) [reply .github/workflows/dockerhub-released-image.yml:32]: In this yml line 45 we will build multi-arch images using `make mp-image-xxx` cmd, but this PR have deleted make mp-image-xxx logic in Makefile, we need to fix this
```
- name: build and publish images
        env: 
          VERSION: ${{ github.ref_name }}
        run: make mp-image-${{ matrix.target }}
```

Comment (ikaven1024) [reply .github/workflows/dockerhub-released-image.yml:32]: `make mp-image-xxx` are merged here, this rule is still work.
```
# Build and push multi-platform image to DockerHub
#
# Example
#   make multi-platform-images
#   make mp-image-karmada-aggregated-apiserver
MP_TARGET=$(addprefix mp-image-, $(TARGETS))
.PHONY: $(MP_TARGET)
$(MP_TARGET):
	set -e;\
	target=$$(echo $(subst mp-image-,,$@));\
	make $$target GOOS=linux GOARCH=amd64;\
	make $$target GOOS=linux GOARCH=arm64;\
	VERSION=$(VERSION) REGISTRY=$(REGISTRY) \
		OUTPUT_TYPE=registry \
		BUILD_PLATFORMS=linux/amd64,linux/arm64 \
		hack/docker.sh $$target
```

https://github.com/ikaven1024/karmada/blob/46ce68f606a182e1105abb149deb7e8be7598eb8/Makefile#L75-L90

Comment (ikaven1024) [reply .github/workflows/dockerhub-released-image.yml:32]: I merged `make xxx`, `make image-xxx`, `make mp-image-xxx`. It reduce the code, and easy to maintain. When new component is added, just append it at 
```
TARGETS := karmada-aggregated-apiserver \
			karmada-controller-manager \
			karmada-scheduler \
			karmada-descheduler \
			karmada-webhook \
			karmada-agent \
			karmada-scheduler-estimator \
			karmada-interpreter-webhook-example
```

Comment (gy95) [reply .github/workflows/dockerhub-released-image.yml:32]: oh, got it

File: .gitignore [hunk 4]
```diff
@@ -10,6 +10,8 @@
 
 # Output of the go coverage tool, specifically when used with LiteIDE
 *.out
+_tmp/
```
Comment (gy95) [inline .gitignore:13]: _tmp is not used for now?

File: .github/workflows/dockerhub-released-image.yml (hunk 3 above)
Comment (gy95) [reply .github/workflows/dockerhub-released-image.yml:32]: by test, these below cmds are ok, make one components
```
make karmada-webhook
make image-karmada-webhook
make mp-image-karmada-webhook
```

and for all components, we now support
```
make all
make images
```
but not support make all mp images.

File: .gitignore (hunk 4 above)
Comment (ikaven1024) [reply .gitignore:13]: `_tmp` is used in `hack/verify-xxx`, I add it handy.

File: .github/workflows/dockerhub-released-image.yml (hunk 3 above)
Comment (gy95) [reply .github/workflows/dockerhub-released-image.yml:32]: /lgtm

Comment (ikaven1024) [reply .github/workflows/dockerhub-released-image.yml:32]: > but not support make all mp images.

`make multi-platform-images` did it. This rule is added before, I didn't change its name.

Comment (gy95) [reply .github/workflows/dockerhub-released-image.yml:32]: oh, sry,

**Comment by ikaven1024** [comment]
@RainbowMango Ready for merging.

**Comment by RainbowMango** [comment]
I'm afraid it'll break [Release workflow](https://github.com/karmada-io/karmada/blob/977725d479515c4c0f0b7a030b3bf5964290a0e4/.github/workflows/release.yml#L28).

**Comment by ikaven1024** [comment]
> I'm afraid it'll break [Release workflow](https://github.com/karmada-io/karmada/blob/977725d479515c4c0f0b7a030b3bf5964290a0e4/.github/workflows/release.yml#L28).

`make kubectl-karmada` will match this rule:
```
# It contains all componets, including karmadactl kubectl-karmada
CMD_TARGET=$(TARGETS) $(CTL_TARGETS)
$(CMD_TARGET): $(SOURCES)
	LDFLAGS=$(LDFLAGS) BUILD_PLATFORMS=$(GOOS)/$(GOARCH) hack/build.sh $@
```

But, sorry for failure of below job, as i change the binary path. I will fix it, and test all the job in [Release workflow](https://github.com/karmada-io/karmada/blob/977725d479515c4c0f0b7a030b3bf5964290a0e4/.github/workflows/release.yml#L28).
https://github.com/karmada-io/karmada/blob/977725d479515c4c0f0b7a030b3bf5964290a0e4/.github/workflows/release.yml#L32-L33

File: hack/build.sh [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+# This script builds go components.
+# You can set the platform to build with BUILD_PLATFORMS, with format: `<os>/<arch>`
+# And binaries will be put in `_output/<os>/<arch>/`
+#
+# Usage:
+#   hack/build.sh <target>
+# Args:
+#   $1:              target to build
+#   BUILD_PLATFORMS: platforms to build. You can set one or more platforms separated by comma.
+#                    e.g.: linux/amd64,linux/amd64
```
Comment (RainbowMango) [inline hack/build.sh:16]:
```suggestion
#                    e.g.: linux/amd64,linux/arm64
```

File: hack/build.sh [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+#   LDFLAGS          pass to the `-ldflags` parameter of go build
+# Examples:
+#   hack/build.sh karmada-aggregated-apiserver
+#   BUILD_PLATFORMS=linux/amd64,linux/amd64 hack/build.sh karmada-aggregated-apiserver
```
Comment (RainbowMango) [inline hack/build.sh:20]:
```suggestion
#   BUILD_PLATFORMS=linux/amd64,linux/arm64 hack/build.sh karmada-aggregated-apiserver
```

File: hack/build.sh [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+# This script builds go components.
+# You can set the platform to build with BUILD_PLATFORMS, with format: `<os>/<arch>`
+# And binaries will be put in `_output/<os>/<arch>/`
+#
+# Usage:
+#   hack/build.sh <target>
+# Args:
+#   $1:              target to build
+#   BUILD_PLATFORMS: platforms to build. You can set one or more platforms separated by comma.
```
Comment (RainbowMango) [inline hack/build.sh:15]: `BUILD_PLATFORMS` belongs to `Environments` not  `Args`, right?

**Review by RainbowMango — COMMENTED** [review]
Just went through the build.sh. Looks pretty good.
The review is still ongoing.

**Comment by ikaven1024** [comment]
Test make & tar, work well
```
➜  make clean
➜  mkdir -p _output/release
➜  
➜  make kubectl-karmada GOOS=linux GOARCH=amd64
➜  make kubectl-karmada GOOS=linux GOARCH=arm64
➜  make kubectl-karmada GOOS=darwin GOARCH=amd64
➜  make kubectl-karmada GOOS=darwin GOARCH=arm64
➜  
➜  os=linux arch=amd64 tar czf _output/release/kubectl-karmada-$os-$arch.tgz LICENSE -C _output/bin/$os/$arch kubectl-karmada
➜  os=linux arch=arm64 tar czf _output/release/kubectl-karmada-$os-$arch.tgz LICENSE -C _output/bin/$os/$arch kubectl-karmada
➜  os=darwin arch=amd64 tar czf _output/release/kubectl-karmada-$os-$arch.tgz LICENSE -C _output/bin/$os/$arch kubectl-karmada
➜  os=darwin arch=arm64 tar czf _output/release/kubectl-karmada-$os-$arch.tgz LICENSE -C _output/bin/$os/$arch kubectl-karmada
➜  
➜  tree _output/
_output/
├── bin
│   ├── darwin
│   │   ├── amd64
│   │   │   └── kubectl-karmada
│   │   └── arm64
│   │       └── kubectl-karmada
│   └── linux
│       ├── amd64
│       │   └── kubectl-karmada
│       └── arm64
│           └── kubectl-karmada
└── release
    ├── kubectl-karmada-darwin-amd64.tgz
    ├── kubectl-karmada-darwin-arm64.tgz
    ├── kubectl-karmada-linux-amd64.tgz
    └── kubectl-karmada-linux-arm64.tgz
```

File: hack/build.sh (hunk 7 above)
Comment (ikaven1024) [reply hack/build.sh:15]: accurately, it is Environments

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -29,13 +29,15 @@ jobs:
...
+      run: mkdir -p _output/release
     - name: Packaging...
-      run: tar czf kubectl-karmada-${{ matrix.goos }}-${{ matrix.goarch }}.tgz kubectl-karmada LICENSE
+      run: tar czf _output/release/kubectl-karmada-${{ matrix.goos }}-${{ matrix.goarch }}.tgz LICENSE -C _output/bin/${{ matrix.goos }}/${{ matrix.goarch }} kubectl-karmada
     - name: Uploading assets...
       if: ${{ !env.ACT }}
       uses: softprops/action-gh-release@v1
       with:
-        files: ./kubectl-karmada-${{ matrix.goos }}-${{ matrix.goarch }}.tgz
+        files: _output/release/kubectl-karmada-${{ matrix.goos }}-${{ matrix.goarch }}.tgz
```
Comment (RainbowMango) [inline .github/workflows/release.yml:40]: By the way, I'm thinking to add a rule in `makefile` to prepare the release assets and let `workflow` just do the upload task. It's pretty easy to test it locally.

How do you think? we can do it after this PR.

**Review by RainbowMango — COMMENTED** [review]
It's fantastic!!! Thanks @ikaven1024 .
Only some questions and nits, otherwise LGTM.

File: Makefile [hunk 9]
```diff
@@ -23,70 +21,79 @@ ifeq ($(VERSION), "")
...
+# 	make all
+#   make karmada-aggregated-apiserver
+#   make karmada-aggregated-apiserver GOOS=linux
+CMD_TARGET=$(TARGETS) $(CTL_TARGETS)
+
+.PHONY: all
+all: $(CMD_TARGET)
+
+.PHONY: $(CMD_TARGET)
+$(CMD_TARGET): $(SOURCES)
```
[the remark this answers is not in the document]
Comment (ikaven1024) [reply Makefile:52]: 1. We assign kubectl-karmada in `CTL_TARGETS`, then assign `CTL_TARGETS` to `CMD_TARGET`
```make
TARGETS := karmada-aggregated-apiserver ...
CTL_TARGETS := karmadactl kubectl-karmada

CMD_TARGET=$(TARGETS) $(CTL_TARGETS)
```
Now  `CMD_TARGET` is `karmada-aggregated-apiserver ... karmadactl kubectl-karmada`.

2. Then 
```
$(CMD_TARGET): $(SOURCES)
	LDFLAGS=$(LDFLAGS) BUILD_PLATFORMS=$(GOOS)/$(GOARCH) hack/build.sh $@
```
will be explained to be:
```make
karmada-aggregated-apiserver ... karmadactl kubectl-karmada: $(SOURCES)
	LDFLAGS=$(LDFLAGS) BUILD_PLATFORMS=$(GOOS)/$(GOARCH) hack/build.sh $@
```

3. So any targets in this list will be matched. And matched target will be set to `$@`.


4. When type `make kubectl-karmada`, match the  `kubectl-karmada` in the list, and assigned to $@.
```make
	LDFLAGS=$(LDFLAGS) BUILD_PLATFORMS=$(GOOS)/$(GOARCH) hack/build.sh $@
```
will be explained to
```make
	LDFLAGS=$(LDFLAGS) BUILD_PLATFORMS=$(GOOS)/$(GOARCH) hack/build.sh kubectl-karmada
```

File: Makefile [hunk 10]
```diff
@@ -23,70 +21,79 @@ ifeq ($(VERSION), "")
...
+# Example:
+#   make image
+#   make image-aggregated-apiserver
+#   make image-aggregated-apiserver GOARCH=arm64
+IMAGE_TARGET=$(addprefix image-, $(TARGETS))
+.PHONY: $(IMAGE_TARGET)
+$(IMAGE_TARGET):
+	set -e;\
+	target=$$(echo $(subst image-,,$@));\
+	make $$target GOOS=linux;\
```
[the remark this answers is not in the document]
Comment (ikaven1024) [reply Makefile:70]: > 1. Why hardcode `GOOS` to `linux` here?  I know some guys run Karmada on Mac.

Here is building image. Docker image is only for linux, none for macOS. People run docker on mac with a linux vm.



> 2\. Why not set `BUILD_PLATFORMS` here?

So far, people set image arch with `GOARCH`, NOT `BUILD_PLATFORMS`, so i didn't change it.
`BUILD_PLATFORMS` is only for script, not for users.

> 3\. Why use `\` to join these commands. I think they are standalone commands.

It seems the mechanism of `Make`, i try to drop the `\`, cannot get the target variable in the last command. I guess `Make`  run each command in a sub process(not verified).

File: Makefile [hunk 11]
```diff
@@ -23,70 +21,79 @@ ifeq ($(VERSION), "")
...
+# Example
+#   make multi-platform-images
+#   make mp-image-karmada-aggregated-apiserver
+MP_TARGET=$(addprefix mp-image-, $(TARGETS))
+.PHONY: $(MP_TARGET)
+$(MP_TARGET):
+	set -e;\
+	target=$$(echo $(subst mp-image-,,$@));\
+	make $$target GOOS=linux GOARCH=amd64;\
+	make $$target GOOS=linux GOARCH=arm64;\
```
[the remark this answers is not in the document]
Comment (ikaven1024) [reply Makefile:86]: As mentioned above.

File: .github/workflows/release.yml (hunk 8 above)
Comment (ikaven1024) [reply .github/workflows/release.yml:40]: LGTM

**Comment by ikaven1024** [comment]
@RainbowMango help please.
When i accept `commit suggestion`, new commit has no `sign-off`. How can i set it?

**Comment by RainbowMango** [comment]
> When i accept `commit suggestion`, new commit has no `sign-off`. How can i set it?

You should [squash](https://github.com/karmada-io/karmada/blob/master/docs/contributors/guide/github-workflow.md#squash-commits) your commits.

File: hack/docker.sh [hunk 12]
```diff
@@ -5,6 +5,23 @@ set -o nounset
...
+# When `OUTPUT_TYPE=docker` is set, `BUILD_PLATFORMS` cannot be set with multi platforms.
+# See: https://github.com/docker/buildx/issues/59
+#
+# Usage:
+#   hack/docker.sh <target>
+# Args:
+#   $1:              target to build
+#   BUILD_PLATFORMS: platforms to build. You can set one or more platforms separated by comma.
+#                    e.g.: linux/amd64,linux/arm64
+#   OUTPUT_TYPE      Destination to save image(`docker`/`registry`/`local,dest=path`, default is `docker`).
```
[the remark this answers is not in the document]
Comment (RainbowMango) [reply hack/docker.sh:19]: Got it. Thanks.

File: Makefile (hunk 10 above)
[the remark this answers is not in the document]
Comment (ikaven1024) [reply Makefile:70]: > Each command is run in a new shell (or at least the effect is as such)

FYI: https://makefiletutorial.com/#command-execution

[the remark this answers is not in the document]
Comment (RainbowMango) [reply Makefile:70]: ok. Thanks.

**Comment by RainbowMango** [comment]
Generally looks good, before moving this forward, I have a question.

It takes [about `8` minutes ](https://github.com/karmada-io/karmada/runs/6381144344?check_suite_focus=true) to get ready for CI. That's awesome! 

I run `hack/local-up-karmada.sh` on my side, but it takes about 13min. It takes about 3~4min before, I'm sure it is the same machine.

@XiShanYongYe-Chang @huone1 Can you help to test how long it will take on your side? I remember you have tested it before.

**Comment by ikaven1024** [comment]
> I run `hack/local-up-karmada.sh` on my side, but it takes about 13min. It takes about 3~4min before, I'm sure it is the same machine.

Below affect it：
1. go cache. If you have build before, then cache will work, we can get the binary few seconds.
2. docker build cache. Pulling base image and `apk add --no-cache ca-certificates` will be skipped.
3. pulling kind image. If local image is updated, this also very fast.

Another tip, If we use the same version of kind with it, k8s images are already in kind image. This can accelerate the start of karmada.
https://github.com/karmada-io/karmada/blob/b2ebfb60361a8ac54fb9e2da9728f08ae5c3a02f/artifacts/deploy/karmada-apiserver.yaml#L67

**Comment by RainbowMango** [comment]
> @RainbowMango you can re-run hack/local-up-karmada.sh for test.

I run several times locally. Almost the same time each time.  :( :(

**Comment by XiShanYongYe-Chang** [comment]
It cost `3m50.004s` on my local site. (4u8g linux x86_64)

**Comment by RainbowMango** [comment]
That must be my problem. I'll debug it then.

/lgtm
/approve

Thanks again @ikaven1024 for your excellent work!

**Comment by RainbowMango** [comment]
The workflow `latest image to DockerHub` is failing on master.
https://github.com/karmada-io/karmada/runs/6383487427?check_suite_focus=true

I've re-triggered once.

**Comment by ikaven1024** [comment]
@RainbowMango I know the reason now. It run `dockerhub-latest-image` job with `swr.ap-southeast-1.myhuaweicloud.com` registry.

**Comment by huone1** [comment]
there is a error when i run hack/local-up-karmada.sh
https://github.com/karmada-io/karmada/blob/da78076e261f796a03c6bb2306cb72664af956e1/hack/docker.sh#L33-L34
image

**Comment by ikaven1024** [comment]
> there is a error when i run hack/local-up-karmada.sh
> 
> https://github.com/karmada-io/karmada/blob/da78076e261f796a03c6bb2306cb72664af956e1/hack/docker.sh#L33-L34

```diff
diff --git a/hack/docker.sh b/hack/docker.sh
index 683c4daf..ed69edd9 100755
--- a/hack/docker.sh
+++ b/hack/docker.sh
@@ -31,7 +31,7 @@ REGISTRY=${REGISTRY:-"swr.ap-southeast-1.myhuaweicloud.com/karmada"}
 VERSION=${VERSION:="unknown"}
 
 function build_images() {
-  local -ra target=$1
+  local target=$1
   local -r output_type=${OUTPUT_TYPE:-docker}
   local platforms="${BUILD_PLATFORMS:-"$(util:host_platform)"}"
 ```

@huone1 could you modify as above and test again. Waiting for you reply.

**Comment by huone1** [comment]
> > there is a error when i run hack/local-up-karmada.sh
> > https://github.com/karmada-io/karmada/blob/da78076e261f796a03c6bb2306cb72664af956e1/hack/docker.sh#L33-L34
> 
> ```diff
> diff --git a/hack/docker.sh b/hack/docker.sh
> index 683c4daf..ed69edd9 100755
> --- a/hack/docker.sh
> +++ b/hack/docker.sh
> @@ -31,7 +31,7 @@ REGISTRY=${REGISTRY:-"swr.ap-southeast-1.myhuaweicloud.com/karmada"}
>  VERSION=${VERSION:="unknown"}
>  
>  function build_images() {
> -  local -ra target=$1
> +  local target=$1
>    local -r output_type=${OUTPUT_TYPE:-docker}
>    local platforms="${BUILD_PLATFORMS:-"$(util:host_platform)"}"
> ```
> 
> @huone1 could you modify as above and test again. Waiting for you reply.

it is OK after deleting the option **-a**

**Comment by ikaven1024** [comment]
@huone1 you delete `-ra` or just `a`?

**Comment by huone1** [comment]
> @huone1 you delete `-ra` or just `a`?

just `a`

**Comment by ikaven1024** [comment]
> just `a`

Thank you for your reply. This `a` is really needless. ok i will delete it.

**Comment by ikaven1024** [comment]
@huone1 Could you show your environment
```
uname -a
bash --version
```
or other useful info。

**Comment by RainbowMango** [comment]
No, It works fine on my side.
```
-bash-4.2# cat /etc/redhat-release 
CentOS Linux release 7.7.1908 (Core)
```

**Comment by huone1** [comment]
image

**Pull request merged**